### PR TITLE
Filter out duplicate payslips from income details

### PR DESCRIPTION
**Pull request by oc-ste**
- [x] Unit tests passing
- [x] Coverage reached
- [x] Acceptance tests passing
- [ ] Reviewed

### CODE REVIEW

File: app/uk/gov/hmrc/tai/service/PaymentsService.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import uk.gov.hmrc.tai.model.domain.{Employment, Payment}
+import uk.gov.hmrc.tai.viewModels.PaymentDetailsViewModel
+
+class PaymentsService @Inject()() {
+  def filterDuplicates(employment: Employment): Seq[PaymentDetailsViewModel] = {
+    val payments = employment.latestAnnualAccount.map(_.payments).getOrElse(Seq.empty[Payment])
+    val paymentsWithoutDuplicates = payments.filterNot(_.duplicate.getOrElse(false))
+
+    paymentsWithoutDuplicates.map(payment =>
+      PaymentDetailsViewModel(payment.date, payment.amount, payment.taxAmount, payment.nationalInsuranceAmount))
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/tai/service/PaymentsService.scala]: Can this not be replaced with `paymentsWithoutDuplicates.map(PaymentDetailsViewModel(_))`

File: app/uk/gov/hmrc/tai/service/PaymentsService.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+package uk.gov.hmrc.tai.service
+
+import javax.inject.Inject
+import uk.gov.hmrc.tai.model.domain.{Employment, Payment}
+import uk.gov.hmrc.tai.viewModels.PaymentDetailsViewModel
+
+class PaymentsService @Inject()() {
+  def filterDuplicates(employment: Employment): Seq[PaymentDetailsViewModel] = {
+    val payments = employment.latestAnnualAccount.map(_.payments).getOrElse(Seq.empty[Payment])
```
Comment (kurtlogan) [inline app/uk/gov/hmrc/tai/service/PaymentsService.scala]: Is payments on `latestAnnualAccount` really an `Option[List[Payment]]`. If it isn't too much work consider refactoring it to be `List[Payment]` instead as all the semantics of an `Option` can be realized with a `List` so `Option[List[_]]` doesn't really make sense

File: test/controllers/YourIncomeCalculationControllerSpec.scala [hunk 3]
```diff
@@ -220,19 +221,28 @@ class YourIncomeCalculationControllerSpec extends PlaySpec
   val personService: PersonService = mock[PersonService]
   val employmentService = mock[EmploymentService]
   val taxAccountService = mock[TaxAccountService]
+  val paymentsService = mock[PaymentsService]
```
Comment (kurtlogan) [inline test/controllers/YourIncomeCalculationControllerSpec.scala]: Do we really need to mock the payments service, if possible use the concrete type

File: app/uk/gov/hmrc/tai/service/PaymentsService.scala (hunk 2 above)
Comment (oc-ste) [reply app/uk/gov/hmrc/tai/service/PaymentsService.scala]: It grabs the latest annual account (the list.max) but if it's empty it's a None, so really it's just an isEmpty wrapper that turns it into a None/AnnualAccount. 

`lazy val latestAnnualAccount: Option[AnnualAccount] = if (annualAccounts.isEmpty) None else Some (annualAccounts.max)`

**Pull request merged**